File: java/src/main/java/com/benchmark/routing/Dijkstra.java
package com.benchmark.routing;

import com.benchmark.graph.Graph;
import com.benchmark.models.Edge;
import com.benchmark.models.RouteResult;

import java.util.*;

/**
 * Dijkstra's single-source shortest path algorithm for multi-objective vehicle routing.
 *
 * <p>Implements a binary min-heap (via {@link PriorityQueue}) with lazy deletion for
 * stale priority queue entries. Operates on pre-computed physical edge costs across
 * three optimization modes: FASTEST, CHEAPEST, and BALANCED.
 *
 * <p>All state is local to the static {@link #solve} method invocation; the class
 * holds no mutable instance state and is therefore trivially thread-safe.
 */
public class Dijkstra {

    /**
     * Priority queue entry pairing a vertex identifier with its accumulated objective cost.
     *
     * <p>Natural ordering compares by cost ascending so that {@link PriorityQueue}
     * acts as a min-heap.
     */
    public static class DijkstraEntry implements Comparable<DijkstraEntry> {
        public double cost;
        public int node;

        public DijkstraEntry(double cost, int node) {
            this.cost = cost;
            this.node = node;
        }

        @Override
        public int compareTo(DijkstraEntry o) {
            return Double.compare(this.cost, o.cost);
        }
    }

    /**
     * Predecessor map entry used for optimal path backtracking.
     */
    public static class PrevEdge {
        public int prevNode;
        public int edgeIndex;

        public PrevEdge(int prevNode, int edgeIndex) {
            this.prevNode = prevNode;
            this.edgeIndex = edgeIndex;
        }
    }

    /**
     * Computes the cost-optimal route using Dijkstra's algorithm.
     *
     * <p>Executes priority-queue-driven edge relaxation over pre-computed physical costs.
     * Stale heap entries are discarded via lazy deletion. Path reconstruction backtracks
     * the predecessor map, accumulating all physical route metrics.
     *
     * @param graph  Pre-built Graph with physics-evaluated edge costs.
     * @param source Origin vertex identifier.
     * @param target Destination vertex identifier.
     * @param mode   Optimization objective: {@code "FASTEST"}, {@code "CHEAPEST"}, or {@code "BALANCED"}.
     * @return A {@link RouteResult} containing the optimal path and metrics,
     *         or {@code null} if the target is unreachable.
     */
    public static RouteResult solve(Graph graph, int source, int target, String mode) {
        long startTime = System.nanoTime();

        if (!graph.nodes.containsKey(source) || !graph.nodes.containsKey(target)) {
            return null;
        }

        if (source == target) {
            RouteResult res = new RouteResult();
            res.source = source;
            res.destination = target;
            res.optimization = mode;
            res.algorithm = "DIJKSTRA";
            res.route = Collections.singletonList(source);
            res.nodesExplored = 1;
            res.executionTimeUs = (System.nanoTime() - startTime) / 1000;
            return res;
        }

        Map<Integer, Double> dist = new HashMap<>();
        Map<Integer, PrevEdge> prev = new HashMap<>();
        PriorityQueue<DijkstraEntry> pq = new PriorityQueue<>();

        dist.put(source, 0.0);
        pq.add(new DijkstraEntry(0.0, source));

        int nodesExplored = 0;
        boolean found = false;

        while (!pq.isEmpty()) {
            DijkstraEntry top = pq.poll();
            int u = top.node;
            double cost = top.cost;

            Double currDist = dist.get(u);
            if (currDist != null && cost > currDist) {
                continue;
            }

            nodesExplored++;

            if (u == target) {
                found = true;
                break;
            }

            List<Graph.HalfEdge> neighbors = graph.adj.get(u);
            if (neighbors != null) {
                for (Graph.HalfEdge halfEdge : neighbors) {
                    int v = halfEdge.targetNode;
                    Edge edge = graph.edges.get(halfEdge.edgeIndex);
                    double weight = graph.getEdgeWeight(edge, mode);
                    double newDist = cost + weight;

                    Double currV = dist.get(v);
                    if (currV == null || newDist < currV) {
                        dist.put(v, newDist);
                        prev.put(v, new PrevEdge(u, halfEdge.edgeIndex));
                        pq.add(new DijkstraEntry(newDist, v));
                    }
                }
            }
        }

        long execUs = (System.nanoTime() - startTime) / 1000;

        if (!found) {
            return null;
        }

        // Backtrack predecessor map to reconstruct the optimal node sequence
        List<Integer> routeNodes = new ArrayList<>();
        List<Integer> routeEdges = new ArrayList<>();
        int curr = target;
        routeNodes.add(curr);

        while (curr != source) {
            PrevEdge p = prev.get(curr);
            routeEdges.add(p.edgeIndex);
            curr = p.prevNode;
            routeNodes.add(curr);
        }

        Collections.reverse(routeNodes);
        Collections.reverse(routeEdges);

        double totDist = 0.0;
        double totTime = 0.0;
        double totFuel = 0.0;
        double totFuelCost = 0.0;
        double totToll = 0.0;

        for (int edgeIdx : routeEdges) {
            Edge e = graph.edges.get(edgeIdx);
            totDist += e.distance;
            totTime += e.travelTimeMin;
            totFuel += e.fuelLiters;
            totFuelCost += e.fuelCost;
            totToll += e.tollCost;
        }

        RouteResult res = new RouteResult();
        res.source = source;
        res.destination = target;
        res.optimization = mode;
        res.algorithm = "DIJKSTRA";
        res.route = routeNodes;
        res.distanceKm = Math.round(totDist * 10000.0) / 10000.0;
        res.travelTimeMin = Math.round(totTime * 10000.0) / 10000.0;
        res.fuelLiters = Math.round(totFuel * 10000.0) / 10000.0;
        res.fuelCost = Math.round(totFuelCost * 10000.0) / 10000.0;
        res.tollCost = Math.round(totToll * 10000.0) / 10000.0;
        res.totalCost = Math.round(dist.get(target) * 10000.0) / 10000.0;
        res.nodesExplored = nodesExplored;
        res.executionTimeUs = execUs;

        return res;
    }
}

File: go/pkg/routing/dijkstra.go
// Package routing provides Dijkstra and A* shortest path implementations
// for multi-objective vehicle route optimization over physical road networks.
//
// All functions are goroutine-safe (graph is read-only after construction)
// and are suitable for concurrent dispatch via goroutine pools.
package routing

import (
	"container/heap"
	"math"
	"route_optimization_go/pkg/graph"
	"route_optimization_go/pkg/models"
	"time"
)

// DijkstraItem is a priority queue entry representing a vertex with its accumulated cost.
type DijkstraItem struct {
	cost  float64
	node  int
	index int
}

type DijkstraPQ []*DijkstraItem

func (pq DijkstraPQ) Len() int           { return len(pq) }
func (pq DijkstraPQ) Less(i, j int) bool { return pq[i].cost < pq[j].cost }
func (pq DijkstraPQ) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}
func (pq *DijkstraPQ) Push(x interface{}) {
	n := len(*pq)
	item := x.(*DijkstraItem)
	item.index = n
	*pq = append(*pq, item)
}
func (pq *DijkstraPQ) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

// PrevEdge stores the predecessor vertex and edge index for path backtracking.
type PrevEdge struct {
	PrevNode  int
	EdgeIndex int
}

// SolveDijkstra computes the cost-optimal route between two vertices using Dijkstra's algorithm.
//
// Implements a binary min-heap (via container/heap) with lazy deletion for stale entries.
// The graph is traversed using pre-computed multi-objective edge costs. Path reconstruction
// backtracks the predecessor map to yield the complete ordered node sequence.
//
// Parameters:
//
//	g      - Pre-built Graph with physics-evaluated edge costs.
//	source - Origin vertex identifier.
//	target - Destination vertex identifier.
//	mode   - Optimization objective: "FASTEST", "CHEAPEST", or "BALANCED".
//
// Returns: Pointer to RouteResult with path and metrics, or nil if unreachable.
func SolveDijkstra(g *graph.Graph, source, target int, mode string) *models.RouteResult {
	startTime := time.Now()

	if _, ok := g.Nodes[source]; !ok {
		return nil
	}
	if _, ok := g.Nodes[target]; !ok {
		return nil
	}

	if source == target {
		return &models.RouteResult{
			Source:          source,
			Destination:     target,
			Optimization:    mode,
			Algorithm:       "DIJKSTRA",
			Route:           []int{source},
			DistanceKm:      0.0,
			TravelTimeMin:   0.0,
			FuelLiters:      0.0,
			FuelCost:        0.0,
			TollCost:        0.0,
			TotalCost:       0.0,
			NodesExplored:   1,
			ExecutionTimeUs: time.Since(startTime).Microseconds(),
		}
	}

	dist := make(map[int]float64)
	prev := make(map[int]PrevEdge)
	pq := &DijkstraPQ{}
	heap.Init(pq)

	dist[source] = 0.0
	heap.Push(pq, &DijkstraItem{cost: 0.0, node: source})

	nodesExplored := 0
	found := false

	for pq.Len() > 0 {
		top := heap.Pop(pq).(*DijkstraItem)
		u := top.node
		cost := top.cost

		if currDist, ok := dist[u]; ok && cost > currDist {
			continue
		}

		nodesExplored++

		if u == target {
			found = true
			break
		}

		for _, halfEdge := range g.Adj[u] {
			v := halfEdge.TargetNode
			edge := &g.Edges[halfEdge.EdgeIndex]
			weight := g.GetEdgeWeight(edge, mode)
			newDist := cost + weight

			if currV, ok := dist[v]; !ok || newDist < currV {
				dist[v] = newDist
				prev[v] = PrevEdge{PrevNode: u, EdgeIndex: halfEdge.EdgeIndex}
				heap.Push(pq, &DijkstraItem{cost: newDist, node: v})
			}
		}
	}

	execUs := time.Since(startTime).Microseconds()

	if !found {
		return nil
	}

	// Backtrack predecessor map to reconstruct the optimal node sequence
	routeNodes := make([]int, 0)
	routeEdges := make([]int, 0)
	curr := target
	routeNodes = append(routeNodes, curr)

	for curr != source {
		p := prev[curr]
		routeEdges = append(routeEdges, p.EdgeIndex)
		curr = p.PrevNode
		routeNodes = append(routeNodes, curr)
	}

	// Reverse route nodes and edges
	for i, j := 0, len(routeNodes)-1; i < j; i, j = i+1, j-1 {
		routeNodes[i], routeNodes[j] = routeNodes[j], routeNodes[i]
	}
	for i, j := 0, len(routeEdges)-1; i < j; i, j = i+1, j-1 {
		routeEdges[i], routeEdges[j] = routeEdges[j], routeEdges[i]
	}

	var totDist, totTime, totFuel, totFuelCost, totToll float64
	for _, idx := range routeEdges {
		e := &g.Edges[idx]
		totDist += e.Distance
		totTime += e.TravelTimeMin
		totFuel += e.FuelLiters
		totFuelCost += e.FuelCost
		totToll += e.TollCost
	}

	return &models.RouteResult{
		Source:          source,
		Destination:     target,
		Optimization:    mode,
		Algorithm:       "DIJKSTRA",
		Route:           routeNodes,
		DistanceKm:      math.Round(totDist*10000.0) / 10000.0,
		TravelTimeMin:   math.Round(totTime*10000.0) / 10000.0,
		FuelLiters:      math.Round(totFuel*10000.0) / 10000.0,
		FuelCost:        math.Round(totFuelCost*10000.0) / 10000.0,
		TollCost:        math.Round(totToll*10000.0) / 10000.0,
		TotalCost:       math.Round(dist[target]*10000.0) / 10000.0,
		NodesExplored:   nodesExplored,
		ExecutionTimeUs: execUs,
	}
}
